core/src/schedule.rs

```diff
@@ -0,0 +1,180 @@
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::ops::RangeFrom;
+
+use bimap::BiBTreeMap;
+use granular_id::{GranularId, UpperBounded};
+use topological_sort::TopologicalSort;
+
+use crate::element::GranId;
+use crate::variables::{VarAccess, Variable};
+use crate::{Context, Element, OutputFormat};
+
+type ScheduleId = usize;
+
+// In short, each element is identified by its "GranId". We can use these Ids in our sorting, and
+// then get the next ID to be evaluated. But when a parent node is evaluated, all its children
+// is invalid and any returned IDs from its children (which is "new") may previously have been
+// assigned to the previous children (which doesn't exist anymore). This means that we need to be
+// able to "remove" IDs, but that isn't possible in TopologicalSort. For this reason, we assign
+// new unique IDs to each GranId, of the type ScheduleId. We have an BiBTreeMap to keep track of
+// this mapping.
+
+pub struct Schedule {
+    dag: TopologicalSort<ScheduleId>,
+    dep_info: HashMap<Variable, Vec<(ScheduleId, VarAccess)>>,
+    id_map: BiBTreeMap<GranId, ScheduleId>,
+    id_iter: RangeFrom<ScheduleId>, // Assume this is infinite
+}
+
+impl Default for Schedule {
+    fn default() -> Self {
+        Self {
+            dag: Default::default(),
+            dep_info: Default::default(),
+            id_map: Default::default(),
+            id_iter: ScheduleId::MIN..,
+        }
+    }
+}
+
+impl Schedule {
+    pub fn is_empty(&self) -> bool {
+        self.dag.is_empty()
+    }
+
+    pub fn pop(&mut self) -> Option<GranId> {
+        // TODO LATER: use pop all and reorder them so modules of the same type are evaluated
+        //  after each other, that way we can keep the same wasm module alive for better performance
+
+        // schedule_id is the ID of the next node to be evaluated
+        while let Some(schedule_id) = self.dag.pop() {
+            // If this isn't in our BiBTreeMap it means that it doesn't exist anymore (it may once
+            // have been a child of an element that is now evaluated). Then we pop the next ID.
+            let Some((gran_id, _)) = self.id_map.remove_by_right(&schedule_id) else {
+                continue;
```

Comment:
Mostly a nit-pick but it can be kind of nice to move the comment to the relevant clause.
```suggestion
            let Some((gran_id, _)) = self.id_map.remove_by_right(&schedule_id) else {
                 // If this isn't in our BiBTreeMap it means that it doesn't exist anymore (it may once
                 // have been a child of an element that is now evaluated). Then we pop the next ID.
                continue;
```